### ISSUE TITLE
Add serial passthrough support on host machines with a serial port

### ISSUE THREAD

**Issue by MoochMcGee** [86Box/86Box#256]
This can be used to, say, debug a Windows kernel running in 86Box on one machine, from another machine.

An example of how to program this on Windows is given here: https://www.codeguru.com/cpp/i-n/network/serialcommunications/article.php/c5425/Serial-Communication-in-Windows.htm

**Comment by andi8086** [comment]
What, this is still open??? I need that feature now XD LOL, this is really basic stuff, I might have a look now into it at least on Unix.

**Comment by andi8086** [comment]
Okay, I need a little help to understand how to add hardware.  I would suggest to add a new device called "serial_passthrough" or something like that and on Linux systems, use a virtual console via /dev/ptmx. However I dont understand how to emulator enumerates available hardware... i.e. what I can set in a config file. Where is the mapping between config file keywords and the devices?

**Comment by ssokolow** [comment]
I'm not familiar with the 86box side of things, but I can advise on design decisions for the Linux side of things as well as what other emulators have done.

First, I'd want some examples of the kind of use-cases you're thinking of.

For example, for me, I'd want something equivalent to DOSbox's support for exposing an emulated serial port as a TCP listening socket, so I can run a testing harness for builds of my DOS creations that have been instrumented to treat COM1 as stderr and pause at various points to wait on COM1 so the harness can reliably take screenshots to compare to known good versions for rendered output tests.

**Comment by ts-korhonen** [comment]
Nice to see people interested in contributing! Please feel welcome to join the [Discord server](https://discord.gg/QXK9XTv) to chat with the other developers :)

**Comment by andi8086** [comment]
:) I am developing a DOS Device driver. And this device driver does things via the serial port. On the other side is a server, which will be a linux program. I use 86Box for testing of the DOS driver. But for this, I must feed through the serial data. My idea is to create a new device, named something like "serial_passthrough". And add a line for PC and up like "serial_passthrough = 1,2" in config, which would then attach the pass_through device to COM1 and COM2 if needed. Then in Linux, I would open /dev/ptmx to get a virtual device and pass the data to and from a new virtual console on read and write. Of course DTS/RTS and speed are not handled, the virtual console more behaves like a FIFO, but this way, also debugging of Kernels should be possible! The virtual terminal can be set to a fixed baud rate, whereas the simulated serial device can keep its baud rate.

In the end it does not matter if reading or writing to/from /dev/pt.... or to/from a socket. In Linux at least. I am not familiar with Mac, so I can only help out with linux. And I really am not eager to do anything on windows. Thus, I would create plat_ functions and use them for the platform specific implementations. On Windows they would be dummy functions on the first run.

I think what I am looking for is in the config.c, where there is a massive amount of strcmp XD

**Comment by andi8086** [comment]
Ah, and by the way... I was a BIOS fetishist as child LOL. I love seeing PCs booting, especially AMI BIOS. And this week I saw you have my favourite BIOS also counting to the exact nuamber 7808K with 8 MB :D This was an amazing feeling to see this in an emulator. And this project somehow attracts me magically. Already got two commits merged. So I hope further things are going to happen :)

**Comment by ssokolow** [comment]
If you don't need the out-of-band signalling stuff like RTS, I'd suggest following DOSBox's lead and using a TCP socket. That way, it's easy to support all platforms with a single implementation, which not only keeps 86box more portable, but makes it easier to allow your host-side testing code to be portable too.

...unless, of course, your primary goal is to have a single implemtnation of your external testing code that works against both 86box and a real PC via a null modem cable. That'd be a justification for 86box exposing something like a named pipe or pty. (or maybe CUSE. It's Linux-specific and I can't remember if it requires root permissions, but it would allow a userland process to create a perfect clone of a kernel-supplied serial port device node.)

**Comment by jriwanek** [comment]
Yup, come and join us on the discord, we're happy for all contributions.

**Comment by ssokolow** [comment]
Me? Aside from me being too busy to learn a new codebase and an utter novice at C, there's just something about Discord that makes it feel like too much of a commitment to join a new server.

**Comment by jriwanek** [comment]
Actually the both of you, but if you're still interested in contributing, we'll be glad to work with you.

**Comment by jriwanek** [comment]
As to examples of what we plan on supporting:
Loopback mode
Loopback Plug
Null Modem
LapLink
Route to real Serial port
Route to system file
Hayes compatible modem 300baud-56k (Will probably implement externally)
Mouse (Existing, but not currently independent of mouse dropdown).
Serial OPL2
Serial Printers

**Comment by ssokolow** [comment]
If you think it'd be useful to be able to `@mention` someone who doesn't have time to keep up with the backlog but can advise on what conventions to follow and what APIs may be available, I can consider it on a day I didn't completely flub getting a good night's sleep.

> Null Modem

Given that "Route to real Serial port" is a separate option, I'm guessing this would be the choice for real-time bi-directional communication with another program outside the emulated machine?

...or is that accomplished by not supporting routing the null modem cable over a LAN and having "Route to system file" support bi-directional communication if pointed at a device node?

**Comment by andi8086** [comment]
@ssokolow Sure, I find the idea with TCP good as well. No Problem with that. The only problem is that I need some kind of adapter then for my real serial server software. So I would like to work on both approaches if we could agree on that.

@jrlwanek Route to real Serial Port...  if that is planned anyway, let me see what I can do!

**Comment by jriwanek** [comment]
Awesome! You can test "real" serial port on windows using com0com, and/or Virtualbox, virtualbox is also a good choice for Linux.

**Comment by jriwanek** [comment]
@ssokolow 
Unfortunately you're looking at my rough notes, so they're not all too clear,

"Given that "Route to real Serial port" is a separate option, I'm guessing this would be the choice for real-time bi-directional communication with another program outside the emulated machine?

...or is that accomplished by not supporting routing the null modem cable over a LAN and having "Route to system file" support bi-directional communication if pointed at a non-serial-port device node?"


People could use the former to use something like com0com or another host side virtual serial port, 
The latter is for cases where you just need a raw dump of what's coming out, or for cases where the external program prefers a file watch, or something like a named pipe, where as a serial over IP server/client would be integrated to 86box itself, much like with dosbox.

**Comment by ssokolow** [comment]
> Unfortunately you're looking at my rough notes, so they're not all too clear,

Yeah. I did get the sense that it was a list of "as the emulated machine sees it" and "as the host machine sees it" items jumbled together.

As for the rest, that selection of options definitely makes sense.

My main concern with serial over IP is ensuring that I don't need to maintain some custom protocol implementation. With DOSBox, I just set `transparent:1` and I've got what I need to write my test instrumentation around, but testing under DOSBox is about as good for compatibility testing as using Wine for Windows testing. The broader the selection of environments I can automate testing under without having a warehouse full of physical hardware, the better.

**Comment by andi8086** [comment]
Already implemented some rudimentary configuration options, which get parsed and stored correctly. Maybe I can provide a first solution draft implementation for review over the weekend :) At least as discussion base for further steps.

**Comment by jriwanek** [comment]
Sounds great, thanks!

**Comment by andi8086** [comment]
How can I trigger a debug build on the commandline with CMAKE? I cannot figure it out... I want to have -g and use gdb...

**Comment by jriwanek** [comment]
cmake -B build 32 --preset=debug -D NEW_DYNAREC=ON -D VNC=OFF for new dynarec,
cmake -B build 32 --preset=debug -D NEW_DYNAREC=OFF -D VNC=OFF for old dynarec
if my memory is correct.

**Comment by andi8086** [comment]
@jriwanek Actually it's -DCMAKE_BUILD_TYPE=debug   :) I am working with a version a little outdated and cant upgrade easily. So I have no --preset.

**Comment by Venomtek** [comment]
Any progress on this?  This is the only thing holding me back from having the complete first PC experience from my childhood.

**Comment by jriwanek** [comment]
It's being worked on slowly on one of my branches, however life is currently getting in the way.

**Comment by ssokolow** [comment]
I just realized that I now have a potential use-case which *would* require a low-level representation to be useful... I'm considering trying to reverse-engineer Laplink's protocol so I could have a native client on my modern Linux box, but Laplink clearly supports doing fancy trickery with the secondary lines since it has a toggle to force a switch from the "7-wire protocol" to the "3-wire protocol" for compatibility.

**Comment by Cacodemon345** [comment]
Serial passthrough is now available.

**Comment by raszpl** [comment]
Would it be possible to link two 86Box instances together using Serial passthrough in Named Pipe mode? I can only select Server right now (4.0 release).
